### PR TITLE
Bug/8/kmod phaseconstraint

### LINKED ISSUE #166: Add phase constraint to kmod
add constraint using phase advance to penalty function in kmod to allow for better measurements with vdM optics

### PR DESCRIPTION
**Pull request by hgarciam**
Fixes #166: Phase constraint added to K-modulation algorithm

### CODE REVIEW

**Review by lmalina — CHANGES_REQUESTED** [review]
It looks that all the new files in tests are not needed, same holds for the madx job files in omc3/kmod/sequences. Also files with just new lines added can be removed from this PR, i.e. checked out from master. 
It would be nice to have test of the phase constraint. 
I will continue the review of the actual method later on.

File: omc3/kmod/analysis.py [hunk 1]
```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+    model_filename = 'twiss_' + kmod_input_params.beam.lower() + '.dat'
+
+    # Position s of BPML
+    if (os.path.exists(os.path.join(f'{kmod_input_params.twiss_model_dir}', f'{model_filename}'))): # distance between actual BPMs if model is provided (more precise)
+        twiss_df = tfs.read(os.path.join(f'{kmod_input_params.twiss_model_dir}', f'{model_filename}'), index='NAME')
+        pos_1L = twiss_df.loc[BPML,'S']
+        pos_1R = twiss_df.loc[BPMR,'S']
+        BPM_distance = np.abs(pos_1R - pos_1L)/2.0
+    else:
+        BPM_distance = 22.965 # Distance between last quadrupole and IP (LSTAR) (less precise)
```
Comment (lmalina) [inline omc3/kmod/analysis.py:289]: is this the same for all IPs? if so leave just a number, if not not model is required or there are also sequence files...
Do we need the model at all?

File: omc3/kmod/analysis.py [hunk 2]
```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+    model_filename = 'twiss_' + kmod_input_params.beam + '.dat'
+
+    # Reset phase advance
+    phase_adv_model = 0.0
+    phase_adv_err = 0.0
+
+    # if measured data exists
+    if os.path.exists(os.path.join(f'{kmod_input_params.meas_directory}',f'getphase{plane.lower()}.out')):
+        #if os.path.exists(os.path.join(f'{kmod_input_params.meas_directory}',f'phase_{plane.lower()}.out')): # this is for python3 phase output
+        phase_adv_model, phase_adv_err = get_phase_from_measurement(kmod_input_params,plane)
```
Comment (lmalina) [inline omc3/kmod/analysis.py:329]: only python 3

File: omc3/kmod/analysis.py [hunk 3]
```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+        pos_1L = twiss_df.loc[BPML,'S']
+        pos_1R = twiss_df.loc[BPMR,'S']
+        BPM_distance = np.abs(pos_1R - pos_1L)/2.0
+    else:
+        BPM_distance = 22.965 # Distance between last quadrupole and IP (LSTAR) (less precise)
+
+    return BPM_distance
+
+def get_phase_from_model(kmod_input_params,plane):        # get phase from twiss model
+    model_filename = 'twiss_' + kmod_input_params.beam + '.dat'
```
Comment (lmalina) [inline omc3/kmod/analysis.py:294]: where do you get twiss_B1/2.dat?

File: omc3/kmod/analysis.py [hunk 4]
```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+    # phase advance constraint
+
+    phase_adv_constraint = phase_constraint(kmod_input_params,plane)
+    BPML,BPMR = get_BPM(kmod_input_params)
+    BPM_distance = get_BPM_distance(kmod_input_params,BPML,BPMR)
+
     for i, s in enumerate(sign):
-        def fun(x): return chi2(x, foc_magnet_df, def_magnet_df, plane, kmod_input_params, s)
+
+        def fun(x): return chi2(x, foc_magnet_df, def_magnet_df, plane, kmod_input_params, s, BPM_distance, phase_adv_constraint, betastar_required)
```
Comment (lmalina) [inline omc3/kmod/analysis.py:392]: 9 params is really too much...

File: omc3/kmod/analysis.py [hunk 5]
```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+
+    # weight given to phase
+    weight = kmod_input_params.phase_weight
+
+    model_filename = 'twiss_' + kmod_input_params.beam + '.dat'
+
+    # Reset phase advance
+    phase_adv_model = 0.0
+    phase_adv_err = 0.0
+
```
Comment (lmalina) [inline omc3/kmod/analysis.py:325]: ???

**Review by lmalina — CHANGES_REQUESTED** [review]
Some more comments and questions...

File: omc3/kmod/analysis.py (hunk 1 above)
Comment (hgarciam) [reply omc3/kmod/analysis.py:289]: Eventually the modulation could be performed in Q2 so the distance between BPMs would change. In this case, the BPM names in the dict above must be changed as well. Need to think how to implement this. In any case, using the sequence and not the model is better.

File: omc3/kmod/analysis.py (hunk 4 above)
Comment (hgarciam) [reply omc3/kmod/analysis.py:392]: reduced to 7.

File: omc3/kmod/analysis.py (hunk 3 above)
Comment (hgarciam) [reply omc3/kmod/analysis.py:293]: twiss_B1/B2 currently must be in --model_dir. We could think a better way.

File: omc3/kmod/analysis.py [hunk 6]
```diff
@@ -247,49 +253,145 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+def phase_constraint(kmod_input_params,plane):
+
+    # weight given to phase
+    weight = kmod_input_params.phase_weight
+
+    model_filename = 'twiss_' + kmod_input_params.beam + '.tfs'
+
+    # if measured data exists
+    #if os.path.exists(os.path.join(f'{kmod_input_params.measurement_dir}',f'getphase{plane.lower()}.out')):
+    if os.path.exists(os.path.join(f'{kmod_input_params.measurement_dir}',f'phase_{plane.lower()}.out')):
```
Comment (lmalina) [inline omc3/kmod/analysis.py:323]: file name root and suffix( which is .tfs) are in omc3/optics_measurements/constants.py

File: omc3/kmod/analysis.py [hunk 7]
```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+
+def get_phase_from_measurement(kmod_input_params,plane):
+    # get measured phase from getphase[x/y].out
+    phase_df = tfs.read( os.path.join(f'{kmod_input_params.meas_directory}',f'getphase{plane.lower()}.out'), index='NAME')
+    ## get measured phase from phase_{x/y}.out
+    # phase_df = tfs.read( os.path.join(f'{kmod_input_params.meas_directory}',f'phase_{plane.lower()}.out'), index='NAME')
+    BPML,BPMR = get_BPM(kmod_input_params)[0], get_BPM(kmod_input_params)[1]
+    phase_adv_model = phase_df.loc[BPML,'PHASE'+plane]
+    phase_adv_err = phase_df.loc[BPML,'STDPH'+plane]
+    # getphase is python2, python3 is phase_x/y
```
[the remark this answers is not in the document]
Comment (lmalina) [reply omc3/kmod/analysis.py:311]: look in omc3/optics_measurements/constants.py

File: omc3/kmod/analysis.py [hunk 8]
```diff
@@ -247,49 +253,145 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+    return phase_adv_model, phase_adv_err
+
+def get_phase_from_measurement(kmod_input_params,plane):
+    # get measured phase from getphase[x/y].out
+    # phase_df = tfs.read( os.path.join(f'{kmod_input_params.measurement_dir}',f'getphase{plane.lower()}.out'), index='NAME')
+    ## get measured phase from phase_{x/y}.out
+    phase_df = tfs.read( os.path.join(f'{kmod_input_params.measurement_dir}',f'phase_{plane.lower()}.out'), index='NAME')
+    BPML,BPMR = get_BPM(kmod_input_params)[0], get_BPM(kmod_input_params)[1]
+    phase_adv_model = phase_df.loc[BPML,'PHASE'+plane]
+    phase_adv_err = phase_df.loc[BPML,'STDPH'+plane]
```
Comment (lmalina) [inline omc3/kmod/analysis.py:310]: f"{ERR}PHASE{plane}"

File: omc3/kmod/analysis.py [hunk 9]
```diff
@@ -247,49 +253,145 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+
+def phase_constraint(kmod_input_params,plane):
+
+    # weight given to phase
+    weight = kmod_input_params.phase_weight
+
+    model_filename = 'twiss_' + kmod_input_params.beam + '.tfs'
+
+    # if measured data exists
+    #if os.path.exists(os.path.join(f'{kmod_input_params.measurement_dir}',f'getphase{plane.lower()}.out')):
```
Comment (lmalina) [inline omc3/kmod/analysis.py:322]: clean up the commented out code

File: omc3/kmod/analysis.py [hunk 10]
```diff
@@ -247,49 +253,145 @@ def return_df(magnet1_df, magnet2_df, plane):
...
+
     for i, s in enumerate(sign):
-        def fun(x): return chi2(x, foc_magnet_df, def_magnet_df, plane, kmod_input_params, s)
+
+        def fun(x): return chi2(x, foc_magnet_df, def_magnet_df, plane, kmod_input_params, s, betastar_required)
         fitresults = scipy.optimize.minimize(fun=fun,
                                              x0=kmod_input_params.betastar_and_waist[plane],
                                              method='nelder-mead',
-                                             tol=1E-22)
+                                             tol=1E-9)
```
Comment (lmalina) [inline omc3/kmod/analysis.py:386]: Is there a reason to relax the tolerance?

**Review by lmalina — CHANGES_REQUESTED** [review]
As discussed during the merge of bug/8/kmod branch, of which this branches of. Please remove the sequence job files (in kmod/sequences/), in tests/inputs/kmod/ remove the simulation job files and consequently the sequence and strength files, as well as the twiss file.

File: omc3/kmod/analysis.py (hunk 10 above)
Comment (hgarciam) [reply omc3/kmod/analysis.py:386]: It was a test I forgot to put back to 1e-22

**Review by lmalina — APPROVED** [review]

**Pull request merged**